### ISSUE TITLE
Vendor drivers (fglrx and nv) fail after swapping AMD card with NVidia card

### ISSUE THREAD

**Issue by egee-irl** [ValveSoftware/SteamOS#427]
I've been using Brewmaster for a little over a month with my AMD R7 360. The fglrx driver packaged with Brewmaster works great, never had any trouble with it.

I recently bought a GTX 750ti and swapped out the R7 360 in my Steam Machine with it. I've discovered now that neither of the vendor drivers work, AMD or Nvidia, at all. Each fail with different results:

1. Plug AMD card in
2. Boot into Brewmaster

Result: System operates as expected.

3. Plug Nvidia card in
4. Boot into Brewmaster

Result: System boots into black screen

5. Plug in AMD card from step 1
6. Boot into Brewmaster

Result: System doesn't appear to boot at all; monitor goes into "sleep mode".

An interesting note is that booting into the recovery mode and using "update-alternates --config glx" and selecting Mesa works fine (after disabling the graphics updater).

Since the system doesn't seem to boot, it's difficult to obtain any debug logs. If I keep pressing ctrl-alt F2 while the AMD card is booting, I get the following screen:

http://i.imgur.com/GncMHHP.jpg

This is a critical issue to me because now I have a non-functional system and reinstalling is not a preferred option.

**Comment by ProfessorKaos64** [comment]
I forget where, but I remember reading you needed to reboot twice. You could also make a backup or just do a restore, and the "preparing hardware drivers" sequence will kick off again, which should set them according to your graphics card. Worth a shot.

**Comment by egee-irl** [comment]
> I remember reading you needed to reboot twice

Nope, definitely not the solution. My system is no longer functional in any capacity, even with the Mesa drivers.

Using the AMD card, no matter which driver I set it to using "update-alternatives --config glx", I get the screen from the image I posted.

The Nvidia card doesn't seem to function at all. Upon logging in I get a black screen and my monitor goes to sleep. Booting into recovery mode also fails.

 I had another, similar issue relating to the way SteamOS switches graphics drivers here - https://github.com/ValveSoftware/SteamOS/issues/389. I am beginning to wonder if Valve tested this functionality at all. Swapping video cards should not cause a catastrophic system failure. I have no choice at this point but to reinstall my system.

I am however curious if the driver "updater" actually works at all. I'm going to use a wiped hard drive and a clean install and swap cards. I am willing to bet that it causes the same issue. I'll post my findings to this report.

**Comment by egee-irl** [comment]
I used the Nvidia card with the Nvidia (*not* Nouveau) and managed to switch over to tty2 before the system locked up. I pulled the kernel log and xorg log file. Reading over the Xorg log it seems as though it is trying to enable both Nvidia and Nouveau drivers. The Nvidia driver fails first, and the Nouveau driver seems to deliver the killing blow to the system.

Kernel log - http://pastebin.com/0NcFLXas
Xorg log - http://pastebin.com/ZJhqZBYs

**Comment by egee-irl** [comment]
Using my living room Steam Machine as my test system, I setup the following test scenario:

* Clean install of SteamOS Brewmaster latest (not the opt-in beta)
* AMD A6 6800k processor
* AMD R7 360 as the initial graphics device
* NVidia GTX 750ti as the final graphics device

The results of this test should help tell us if SteamOS can successfully switch between graphics card vendors without system instability.

Last night, I ran the following test case:

1. Using a clean system (wiped hdd, etc), install SteamOS Brewmaster using ISO provided by Valve
2. Upon install, use the system to establish baseline functionality expectations (Use browser, download updates, etc).
3. Shut system down and install Nvidia graphics card
4. Start system and boot into SteamOS
5. Use system using methodology established in step 2.
6. Shut system down and install AMD graphics card
7. Start system and boot into SteamOS
8. Use system using methodology established in step 2.

Expected Result: 

System should operate normally regardless of the graphics card used.

Actual Results:

System struggled to boot with Nvidia card and required two reboots before fully loading SteamOS. Re-installing the AMD card caused the system to fail to boot entirely. 

Details:

I used the AMD card while installing SteamOS. It booted just fine after completing the install. I used the browser to access the internet, viewed YouTube, checked for updates, etc to verify the system was stable enough to proceed with the test.

After installing the Nvidia card -
The system booted to the SteamOS logo but quickly alternated between a blank black screen and a black screen with a large white cursor. I allowed this to continue for about 15 minutes before I rebooted the system. After rebooting, the system once again reached the SteamOS logo but this time it eventually made it into the SteamOS UI. The sound did not work until I reconfigured it. After fixing the sound issue, the system ran flawlessly.

After re-installing the original AMD card -
The system never even reached the SteamOS logo. As soon as the bios screen was finished, the screen went black and never recovered. I tried rebooting about 4 times before I gave up.

Conclusion:

It appears that something is fundamentally broken with the way SteamOS handles changing or switching graphics drivers for different vendors. I suggest that Valve's internal QA team test this functionality in more detail because there's little more I can do with this issue at this point.
